grafana/grafana
# InfluxDB: Fix sending retention policy with the backend request

**What is this feature?**

This adds a retention policy to the request going from backend to the influxdb. This has to be handled separately for each query. 

Why do we need that big change?
See: https://github.com/grafana/grafana/pull/72467#discussion_r1277576036
Previously we merge all the queries into one big query string and send it to influxdb. And parse the response of it. 
Since we have to send a retention policy for each query separately all queries and their responses should be handled separately. That needs some refactoring.

I did some benchmarking and this code is slightly (but really slightly) better. We now allocate 2 fewer objects 🎉 

```
~/Documents/workspace/grafana ismail/add-retention-policy-to-the-backend-request*
❯ go test -benchmem -run=^$ -memprofile memprofile_old.out -count=10 -bench ^BenchmarkParseJson$ github.com/grafana/grafana/pkg/tsdb/influxdb
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/tsdb/influxdb
BenchmarkParseJson-10               8774            135690 ns/op          156870 B/op       2802 allocs/op
BenchmarkParseJson-10               8672            135430 ns/op          156870 B/op       2802 allocs/op
BenchmarkParseJson-10               8640            135534 ns/op          156870 B/op       2802 allocs/op
BenchmarkParseJson-10               8977            135587 ns/op          156870 B/op       2802 allocs/op
BenchmarkParseJson-10               8593            135506 ns/op          156870 B/op       2802 allocs/op
BenchmarkParseJson-10               8734            135524 ns/op          156871 B/op       2802 allocs/op
BenchmarkParseJson-10               8719            135589 ns/op          156870 B/op       2802 allocs/op
BenchmarkParseJson-10               8752            135542 ns/op          156870 B/op       2802 allocs/op
BenchmarkParseJson-10               8766            135547 ns/op          156870 B/op       2802 allocs/op
BenchmarkParseJson-10               8924            135795 ns/op          156870 B/op       2802 allocs/op
PASS
ok      github.com/grafana/grafana/pkg/tsdb/influxdb    13.222s


~/Documents/workspace/grafana ismail/add-retention-policy-to-the-backend-request* 13s
❯ go test -benchmem -run=^$ -memprofile memprofile_new.out -count=10 -bench ^BenchmarkParseJson$ github.com/grafana/grafana/pkg/tsdb/influxdb/influxql
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/tsdb/influxdb/influxql
BenchmarkParseJson-10               8047            148094 ns/op          156323 B/op       2800 allocs/op
BenchmarkParseJson-10               8350            147282 ns/op          156323 B/op       2800 allocs/op
BenchmarkParseJson-10               8107            147170 ns/op          156321 B/op       2800 allocs/op
BenchmarkParseJson-10               8335            147373 ns/op          156323 B/op       2800 allocs/op
BenchmarkParseJson-10               8006            147547 ns/op          156322 B/op       2800 allocs/op
BenchmarkParseJson-10               8149            147546 ns/op          156322 B/op       2800 allocs/op
BenchmarkParseJson-10               8062            147817 ns/op          156322 B/op       2800 allocs/op
BenchmarkParseJson-10               8121            146934 ns/op          156323 B/op       2800 allocs/op
BenchmarkParseJson-10               8198            147392 ns/op          156323 B/op       2800 allocs/op
BenchmarkParseJson-10               8346            148163 ns/op          156321 B/op       2800 allocs/op
PASS
ok      github.com/grafana/grafana/pkg/tsdb/influxdb/influxql   13.398s
```


Part of https://github.com/grafana/grafana/issues/65045

**Who is this feature for?**

InfluxDB Backend mode users

**How to test?**:
- Enable `influxdbBackendMigration` feature toggle
- Run frontend and backend as usual
- Set up mitmproxy or some other proxy to see backend communication 
  - `mitmproxy --mode=reverse:http://127.0.0.1:8087 --listen-host=127.0.0.1 --listen-port=9101 --set console_mouse=false`
- Create an InfluxDB `1.x` or `2.x` datasource. URL is `http://localhost:9101`
- Go to explore and run a simple influx query with a retention policy named something not `default` 
  - Why? -> https://github.com/grafana/grafana/pull/72467
-  See on proxy there is no `rp` parameter added to the query
- Switch to this branch and try again
- See `rp` parameter added to the query
  - `POST` request has it as encoded
  - `GET` request has it as urlencoded 
  - See `influxql.go > createRequest` for details


**Special notes for your reviewer:**

Please check that:
